### DISCUSSION TITLE
Is there anything "missing" from this?

### DISCUSSION THREAD

**Discussion by The-Futurist** [Microsoft/LLVMSharp#222] [Q&A] [answered]
I'm planning a compiler project and LLVM has come up as the most sensible approach to creating the back end. I want to build the entire system in C#. The front end is not a problem, Antlr4 generates C# parsers so that's good but I don't know much at all about LLVM or if there's anything important missing from LLVMSharp that could cause a problem.

Also is this up to date, does it reflect a reasonably recent version of the actual LLVM API?

**Comment by tannergooding** [comment, marked as answer]
> there's anything important missing from LLVMSharp that could cause a problem.

LLVMSharp is simply bindings over the official LLVM-C library provided by LLVM themselves. LLVM-C is not as feature complete or comprehensive as the main LLVM API and while the vast majority of important features are there, there are bits "missing".

ClangSharp makes up for this by providing libClangSharp to fill in the gaps. I have plans on providing a similar `libLLVMSharp` but haven't had time to actually implement that yet.

> Also is this up to date, does it reflect a reasonably recent version of the actual LLVM API?

It's currently v16.0.0 while the latest official version of LLVM is v18.0.0 and was released last month. It doesn't take much to update since the bindings are just directly generated, I just haven't had time to update as indicated above.

LLVM ships a new major version every 6 months or so and that generally includes binary breaking changes, so many devs opt to not stay on the bleeding edge.

**Reply by ds5678** [reply]
> I have plans on providing a similar libLLVMSharp but haven't had time to actually implement that yet.

I'd be willing to help out with this. I've noticed that when parsing IR, there doesn't seem to be an easy way to get a function's return type. I think it's only exposed in the C++ library.
